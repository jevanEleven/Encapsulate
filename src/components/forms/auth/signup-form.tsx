"use client";

import {
    Form,
    FormControl,
    FormField,
    FormItem,
    FormLabel,
    FormMessage,
} from "@/components/ui/form";
import {
    emailSchema,
    passwordSchema,
    usernameSchema,
} from "@/lib/validations/auth";

import { AuthError } from "@supabase/supabase-js";
import { Button } from "@/components/ui/button";
import { Icons } from "@/components/icons";
import { Input } from "@/components/ui/input";
import { PasswordInput } from "@/components/password-input";
import React from "react";
import getSupabaseClient from "@/db/supabase-client";
import { useForm } from "react-hook-form";
import { useMutation } from "react-query";
import { useRouter } from "next/navigation";
import { z } from "zod";
import { zodResolver } from "@hookform/resolvers/zod";

const formSchema = z.object({
    username: usernameSchema,
    email: emailSchema,
    password: passwordSchema,
});

type Inputs = z.infer<typeof formSchema>;

const dbClient = getSupabaseClient();

export function SignUpForm() {
    const router = useRouter();
    const [errorMessage, setErrorMessage] = React.useState<string>("");

    const form = useForm<Inputs>({
        resolver: zodResolver(formSchema),
        defaultValues: {
            username: "",
            email: "",
            password: "",
        },
    });

    const mutation = useMutation(
        async (data: Inputs) => {
            const { data: resSignUp, error: errorSignUp } =
                await dbClient.auth.signUp({
                    ...data,
                    options: {
                        data: {
                            username: data.username,
                        },
                    },
                });

            if (errorSignUp) {
                throw new AuthError(errorSignUp.message);
            }

            const { user } = resSignUp;
            if (user) {
                await dbClient.from("users").insert({
                    name: user.user_metadata.username,
                    email: user.email as string,
                    id: user.id,
                });
            }
        },
        { retry: false }
    );

    async function onSubmit(data: Inputs) {
        mutation.mutate(data, {
            onSuccess: () => {
                router.push(
                    `/signup/verify-email?email=${data.email}&resend=false`
                );
            },
            onError: (e: any) => {
                if (e instanceof AuthError) {
                    setErrorMessage(e.message);
                } else {
                    setErrorMessage("Unknown Error.");
                }
            },
        });
    }

    return (
        <Form {...form}>
            <form
                className="grid gap-4"
                onSubmit={(...args) =>
                    void form.handleSubmit(onSubmit)(...args)
                }
            >
                <FormField
                    control={form.control}
                    name="username"
                    render={({ field }) => (
                        <FormItem>
                            <FormLabel>Username</FormLabel>
                            <FormControl>
                                <Input placeholder="username" {...field} />
                            </FormControl>
                            <FormMessage />
                        </FormItem>
                    )}
                />
                <FormField
                    control={form.control}
                    name="email"
                    render={({ field }) => (
                        <FormItem>
                            <FormLabel>Email</FormLabel>
                            <FormControl>
                                <Input
                                    placeholder="someone@example.com"
                                    {...field}
                                />
                            </FormControl>
                            <FormMessage />
                        </FormItem>
                    )}
                />
                <FormField
                    control={form.control}
                    name="password"
                    render={({ field }) => (
                        <FormItem>
                            <FormLabel>Password</FormLabel>
                            <FormControl>
                                <PasswordInput
                                    placeholder="**********"
                                    {...field}
                                />
                            </FormControl>
                            <FormMessage />
                        </FormItem>
                    )}
                />
                <Button disabled={mutation.isLoading}>
                    {mutation.isLoading && (
                        <Icons.spinner
                            className="mr-2 size-4 animate-spin"
                            aria-hidden="true"
                        />
                    )}
                    Continue
                    <span className="sr-only">
                        Continue to email verification page
                    </span>
                </Button>
                {mutation.isError && <span>{errorMessage}</span>}
            </form>
        </Form>
    );
}
