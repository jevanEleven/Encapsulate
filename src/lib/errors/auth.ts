class APIError extends Error {
    statusCode: number;

    constructor(message: string, statusCode: number) {
        super(message);
        this.statusCode = statusCode;
        Object.setPrototypeOf(this, APIError.prototype);
    }
}

class APIAuthError extends APIError {
    constructor(message: string, statusCode: number = 401) {
        super(message, statusCode);
        Object.setPrototypeOf(this, APIAuthError.prototype);
    }
}

export { APIError, APIAuthError };
