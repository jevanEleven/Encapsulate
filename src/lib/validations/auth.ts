import BadWordsFilter from "bad-words";
import { z } from "zod";

const usernameRegex = /^[A-Za-z0-9._]{4,20}$/;

const usernameSchema = z
    .string()
    .min(4, { message: "Username must be at least 4 characters long." })
    .max(20, { message: "Username must be no more than 20 characters long." })
    .refine((val) => usernameRegex.test(val), {
        message:
            "Username can only contain alphanumeric characters, '.', and '_'.",
    })
    .refine(
        (val) =>
            !val.startsWith(".") &&
            !val.startsWith("_") &&
            !val.endsWith(".") &&
            !val.endsWith("_"),
        {
            message: "Username cannot start or end with '.' or '_'.",
        }
    )
    .refine((val) => !/[_\.]{2,}/.test(val), {
        message: "Username cannot have consecutive '.' or '_'.",
    })
    .refine((val) => !new BadWordsFilter().isProfane(val), {
        message: "Username failed profanity check.",
    });

const emailSchema = z.string().email();

const passwordSchema = z
    .string()
    .max(64, {
        message: "Password must be at most 64 characters long.",
    })
    .min(8, {
        message: "Password must be at least 8 characters long.",
    })
    .refine((val) => /[a-z]/.test(val), {
        message: "Password must include at least one lowercase letter.",
    })
    .refine((val) => /[A-Z]/.test(val), {
        message: "Password must include at least one uppercase letter.",
    })
    .refine((val) => /[0-9]/.test(val), {
        message: "Password must include at least one number.",
    })
    .refine((val) => /[\^$*.[\]{}()?"!@#%&/,><':;|_~`\\]/.test(val), {
        message: "Password must include at least one special character.",
    });

const credentialsSchema = z.object({
    email: emailSchema,
    password: passwordSchema,
});

const verificationCodeRegex = new RegExp("^[0-9]{6}$");

const verificationCodeSchema = z
    .string()
    .max(6, { message: "Code must be at least 6 characters." })
    .min(6, { message: "Code must be at most 6 characters." })
    .refine((val) => verificationCodeRegex.test(val), {
        message: "Code only contains numbers.",
    });

export {
    usernameSchema,
    emailSchema,
    passwordSchema,
    credentialsSchema,
    verificationCodeSchema,
};
