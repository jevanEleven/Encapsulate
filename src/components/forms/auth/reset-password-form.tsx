"use client";

import { AuthError, EmailOtpType } from "@supabase/supabase-js";
import {
    Form,
    FormControl,
    FormField,
    FormItem,
    FormLabel,
    FormMessage,
} from "@/components/ui/form";
import { useRouter, useSearchParams } from "next/navigation";

import { Button } from "../../ui/button";
import { Icons } from "../../icons";
import { PasswordInput } from "@/components/password-input";
import React from "react";
import getSupabaseClient from "@/db/supabase-client";
import { passwordSchema } from "@/lib/validations/auth";
import { useForm } from "react-hook-form";
import { useMutation } from "react-query";
import { z } from "zod";
import { zodResolver } from "@hookform/resolvers/zod";

const formSchema = z
    .object({
        password: passwordSchema,
        confirmPassword: passwordSchema,
    })
    .refine((data) => data.password === data.confirmPassword, {
        message: "Passwords do not match",
        path: ["confirmPassword"],
    });

type Inputs = z.infer<typeof formSchema>;

const dbClient = getSupabaseClient();

export function ResetPasswordForm() {
    const router = useRouter();
    const searchParams = useSearchParams();
    const [errorMessage, setErrorMessage] = React.useState<string>("");

    const form = useForm<Inputs>({
        resolver: zodResolver(formSchema),
        defaultValues: {
            password: "",
            confirmPassword: "",
        },
    });

    const mutation = useMutation(
        async (data: Inputs) => {
            const token_hash = searchParams.get("token_hash");
            const type = searchParams.get("type") as EmailOtpType | null;

            if (token_hash && type) {
                const { error: authError } = await dbClient.auth.verifyOtp({
                    type,
                    token_hash,
                });

                if (authError) {
                    throw new AuthError(authError.message);
                }

                const { data: resetData, error: resetError } =
                    await dbClient.auth.updateUser({
                        password: data.password,
                    });

                if (resetError) {
                    throw new AuthError(resetError.message);
                }
            }
        },
        { retry: false }
    );

    async function onSubmit(data: Inputs) {
        mutation.mutate(data, {
            onSuccess: () => {
                router.push("/");
            },
            onError: (e: any) => {
                if (e instanceof AuthError) {
                    setErrorMessage(e.message);
                } else {
                    setErrorMessage("Unknown Error.");
                }
            },
        });
    }

    return (
        <Form {...form}>
            <form
                className="grid gap-4"
                onSubmit={(...args) =>
                    void form.handleSubmit(onSubmit)(...args)
                }
            >
                <FormField
                    control={form.control}
                    name="password"
                    render={({ field }) => (
                        <FormItem>
                            <FormLabel>New Password</FormLabel>
                            <FormControl>
                                <PasswordInput
                                    placeholder="**********"
                                    {...field}
                                />
                            </FormControl>
                            <FormMessage />
                        </FormItem>
                    )}
                />
                <FormField
                    control={form.control}
                    name="confirmPassword"
                    render={({ field }) => (
                        <FormItem>
                            <FormLabel>Confirm New Password</FormLabel>
                            <FormControl>
                                <PasswordInput
                                    placeholder="**********"
                                    {...field}
                                />
                            </FormControl>
                            <FormMessage />
                        </FormItem>
                    )}
                />
                <Button disabled={mutation.isLoading}>
                    {mutation.isLoading && (
                        <Icons.spinner
                            className="mr-2 size-4 animate-spin"
                            aria-hidden="true"
                        />
                    )}
                    Reset Password
                    <span className="sr-only">Reset password</span>
                </Button>
                {mutation.isError && <span>{errorMessage}</span>}
            </form>
        </Form>
    );
}
