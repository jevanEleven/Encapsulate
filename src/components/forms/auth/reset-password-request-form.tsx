"use client";

import {
    Form,
    FormControl,
    FormField,
    FormItem,
    FormLabel,
    FormMessage,
} from "@/components/ui/form";

import { Button } from "@/components/ui/button";
import { Icons } from "@/components/icons";
import { Input } from "@/components/ui/input";
import React from "react";
import { emailSchema } from "@/lib/validations/auth";
import getSupabaseClient from "@/db/supabase-client";
import { useForm } from "react-hook-form";
import { useMutation } from "react-query";
import { z } from "zod";
import { zodResolver } from "@hookform/resolvers/zod";

const formSchema = z.object({
    email: emailSchema,
});

type Inputs = z.infer<typeof formSchema>;

const dbClient = getSupabaseClient();

export function ResetPasswordRequestForm() {
    const form = useForm<Inputs>({
        resolver: zodResolver(formSchema),
        defaultValues: {
            email: "",
        },
    });

    const mutation = useMutation(
        async (data: Inputs) => {
            await dbClient.auth.resetPasswordForEmail(data.email);
        },
        { retry: false }
    );

    async function onSubmit(data: Inputs) {
        mutation.mutate(data);
    }

    return (
        <Form {...form}>
            <form
                className="grid gap-4"
                onSubmit={(...args) =>
                    void form.handleSubmit(onSubmit)(...args)
                }
            >
                <FormField
                    control={form.control}
                    name="email"
                    render={({ field }) => (
                        <FormItem>
                            <FormLabel>Email</FormLabel>
                            <FormControl>
                                <Input
                                    placeholder="someone@example.com"
                                    {...field}
                                />
                            </FormControl>
                            <FormMessage />
                        </FormItem>
                    )}
                />
                <Button disabled={mutation.isLoading || mutation.isSuccess}>
                    {mutation.isLoading && (
                        <Icons.spinner
                            className="mr-2 size-4 animate-spin"
                            aria-hidden="true"
                        />
                    )}
                    Continue
                    <span className="sr-only">
                        Continue to email verification page
                    </span>
                </Button>
                {mutation.isSuccess && (
                    <span>
                        If an account with this email address exists
                        <br />
                        we sent you an email with instructions
                        <br />
                        on resetting your password.
                    </span>
                )}
            </form>
        </Form>
    );
}
