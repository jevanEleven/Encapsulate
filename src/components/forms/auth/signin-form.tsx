"use client";

import {
    Form,
    FormControl,
    FormField,
    FormItem,
    FormLabel,
    FormMessage,
} from "@/components/ui/form";
import { emailSchema, passwordSchema } from "@/lib/validations/auth";
import { useRouter, useSearchParams } from "next/navigation";

import { AuthError } from "@supabase/supabase-js";
import { Button } from "@/components/ui/button";
import { Icons } from "@/components/icons";
import { Input } from "@/components/ui/input";
import { PasswordInput } from "@/components/password-input";
import React from "react";
import getSupabaseClient from "@/db/supabase-client";
import { useForm } from "react-hook-form";
import { useMutation } from "react-query";
import { z } from "zod";
import { zodResolver } from "@hookform/resolvers/zod";

const formSchema = z.object({
    email: emailSchema,
    password: passwordSchema,
});

type Inputs = z.infer<typeof formSchema>;

const dbClient = getSupabaseClient();

export function SignInForm() {
    const router = useRouter();
    const params = useSearchParams();
    const [errorMessage, setErrorMessage] = React.useState<string>("");

    // react-hook-form
    const form = useForm<Inputs>({
        resolver: zodResolver(formSchema),
        defaultValues: {
            email: "",
            password: "",
        },
    });

    const mutation = useMutation(
        async (data: Inputs) => {
            const { data: resSignIn, error: errorSignIn } =
                await dbClient.auth.signInWithPassword(data);

            if (errorSignIn) {
                throw new AuthError(errorSignIn.message);
            }
        },
        { retry: false }
    );

    async function onSubmit(data: Inputs) {
        mutation.mutate(data, {
            onSuccess: () => {
                router.push(params.get("callback") || "/");
            },
            onError: (e: any) => {
                if (e instanceof AuthError) {
                    setErrorMessage(e.message);

                    if (e.message === "Email not confirmed") {
                        router.push(
                            `/signup/verify-email?email=${data.email}&resend=true`
                        );
                    }
                } else {
                    setErrorMessage("Unknown Error.");
                }
            },
        });
    }

    return (
        <Form {...form}>
            <form
                className="grid gap-4"
                onSubmit={(...args) =>
                    void form.handleSubmit(onSubmit)(...args)
                }
            >
                <FormField
                    control={form.control}
                    name="email"
                    render={({ field }) => (
                        <FormItem>
                            <FormLabel>Email</FormLabel>
                            <FormControl>
                                <Input
                                    placeholder="someone@example.com"
                                    {...field}
                                />
                            </FormControl>
                            <FormMessage />
                        </FormItem>
                    )}
                />
                <FormField
                    control={form.control}
                    name="password"
                    render={({ field }) => (
                        <FormItem>
                            <FormLabel>Password</FormLabel>
                            <FormControl>
                                <PasswordInput
                                    placeholder="**********"
                                    {...field}
                                />
                            </FormControl>
                            <FormMessage />
                        </FormItem>
                    )}
                />
                <Button disabled={mutation.isLoading}>
                    {mutation.isLoading && (
                        <Icons.spinner
                            className="mr-2 size-4 animate-spin"
                            aria-hidden="true"
                        />
                    )}
                    Sign In
                    <span className="sr-only">Sign in</span>
                </Button>
                {mutation.isError && <span>{errorMessage}</span>}
            </form>
        </Form>
    );
}
