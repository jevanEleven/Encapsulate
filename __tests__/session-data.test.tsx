import React from 'react';
import { render } from '@testing-library/react';
import { useRouter, useSearchParams } from 'next/navigation';
import { useMutation } from 'react-query';
import { ResetPasswordForm } from '@/components/forms/auth/reset-password-form';

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
  useSearchParams: jest.fn()
}));

jest.mock('react-query', () => ({
  useMutation: jest.fn(() => ({
    mutate: jest.fn(),
    isLoading: false,
    isError: false,
    error: null
  }))
}));

describe('ResetPasswordForm', () => {
  it('renders correctly', () => {
    // Mock router and searchParams hooks
    (useRouter as jest.Mock).mockReturnValue({
      push: jest.fn()
    });
    (useSearchParams as jest.Mock).mockReturnValue({
      get: jest.fn((param) => {
        if (param === 'token_hash') return 'some_token_hash';
        if (param === 'type') return 'recovery';
      })
    });

    const { asFragment } = render(<ResetPasswordForm />);
    expect(asFragment()).toMatchSnapshot();
  });
});
