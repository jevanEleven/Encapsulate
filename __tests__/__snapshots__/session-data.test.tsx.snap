// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPasswordForm renders correctly 1`] = `
<DocumentFragment>
  <form
    class="grid gap-4"
  >
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r0:-form-item"
      >
        New Password
      </label>
      <div
        class="relative"
      >
        <input
          aria-describedby=":r0:-form-item-description"
          aria-invalid="false"
          class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 pr-10"
          id=":r0:-form-item"
          name="password"
          placeholder="**********"
          type="password"
          value=""
        />
        <button
          class="inline-flex items-center justify-center font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 hover:text-accent-foreground rounded-md text-xs absolute right-0 top-0 h-full px-3 py-1 hover:bg-transparent"
          disabled=""
          type="button"
        >
          <svg
            aria-hidden="true"
            class="size-4"
            fill="none"
            height="15"
            viewBox="0 0 15 15"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M7.5 11C4.80285 11 2.52952 9.62184 1.09622 7.50001C2.52952 5.37816 4.80285 4 7.5 4C10.1971 4 12.4705 5.37816 13.9038 7.50001C12.4705 9.62183 10.1971 11 7.5 11ZM7.5 3C4.30786 3 1.65639 4.70638 0.0760002 7.23501C-0.0253338 7.39715 -0.0253334 7.60288 0.0760014 7.76501C1.65639 10.2936 4.30786 12 7.5 12C10.6921 12 13.3436 10.2936 14.924 7.76501C15.0253 7.60288 15.0253 7.39715 14.924 7.23501C13.3436 4.70638 10.6921 3 7.5 3ZM7.5 9.5C8.60457 9.5 9.5 8.60457 9.5 7.5C9.5 6.39543 8.60457 5.5 7.5 5.5C6.39543 5.5 5.5 6.39543 5.5 7.5C5.5 8.60457 6.39543 9.5 7.5 9.5Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
          <span
            class="sr-only"
          >
            Show password
          </span>
        </button>
      </div>
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r1:-form-item"
      >
        Confirm New Password
      </label>
      <div
        class="relative"
      >
        <input
          aria-describedby=":r1:-form-item-description"
          aria-invalid="false"
          class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 pr-10"
          id=":r1:-form-item"
          name="confirmPassword"
          placeholder="**********"
          type="password"
          value=""
        />
        <button
          class="inline-flex items-center justify-center font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 hover:text-accent-foreground rounded-md text-xs absolute right-0 top-0 h-full px-3 py-1 hover:bg-transparent"
          disabled=""
          type="button"
        >
          <svg
            aria-hidden="true"
            class="size-4"
            fill="none"
            height="15"
            viewBox="0 0 15 15"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M7.5 11C4.80285 11 2.52952 9.62184 1.09622 7.50001C2.52952 5.37816 4.80285 4 7.5 4C10.1971 4 12.4705 5.37816 13.9038 7.50001C12.4705 9.62183 10.1971 11 7.5 11ZM7.5 3C4.30786 3 1.65639 4.70638 0.0760002 7.23501C-0.0253338 7.39715 -0.0253334 7.60288 0.0760014 7.76501C1.65639 10.2936 4.30786 12 7.5 12C10.6921 12 13.3436 10.2936 14.924 7.76501C15.0253 7.60288 15.0253 7.39715 14.924 7.23501C13.3436 4.70638 10.6921 3 7.5 3ZM7.5 9.5C8.60457 9.5 9.5 8.60457 9.5 7.5C9.5 6.39543 8.60457 5.5 7.5 5.5C6.39543 5.5 5.5 6.39543 5.5 7.5C5.5 8.60457 6.39543 9.5 7.5 9.5Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
          <span
            class="sr-only"
          >
            Show password
          </span>
        </button>
      </div>
    </div>
    <button
      class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
    >
      Reset Password
      <span
        class="sr-only"
      >
        Reset password
      </span>
    </button>
  </form>
</DocumentFragment>
`;
